src/prof_recent.c

```diff
@@ -0,0 +1,243 @@
+		node = prof_recent_alloc_head;
+		prof_recent_alloc_head = node->next;
+		if (prof_recent_alloc_head == NULL) {
+			prof_recent_alloc_tail = NULL;
+		}
+		--prof_recent_alloc_count;
+		head_tctx = node->tctx;
+	} else {
+		/* Otherwise allocate a new node. */
+		node = (prof_recent_alloc_t *)iallocztm(tsd_tsdn(tsd),
+		    sizeof(prof_recent_alloc_t),
+		    sz_size2index(sizeof(prof_recent_alloc_t)),
+		    false, NULL, true, arena_get(TSDN_NULL, 0, true), true);
+		if (unlikely(node == NULL)) {
+			malloc_mutex_unlock(tsd_tsdn(tsd),
+			    &prof_recent_alloc_mtx);
+			/*
+			 * The increment of ref_count in prof_recent_alloc_fast
+			 * was a mistake - fix it.
+			 */
+			decrement_ref_count(tsd, tctx);
+			return true;
+		}
+		head_tctx = NULL;
+	}
+
+	/* Ready to add the new node. */
+	assert(prof_recent_alloc_max == -1 || (prof_recent_alloc_max > 0 &&
+	    prof_recent_alloc_count < prof_recent_alloc_max));
+	assert(node != NULL);
+
+	/* Fill in node content. */
+	nstime_t alloc_time = prof_alloc_time_get(tsd_tsdn(tsd), ptr, NULL);
+	node->alloc_time = nstime_ns(&alloc_time);
+	node->usize = usize;
+	node->tctx = tctx;
+
+	/* Append to the list */
+	node->next = NULL;
+	if (prof_recent_alloc_head == NULL) {
+		assert(prof_recent_alloc_tail == NULL);
+		assert(prof_recent_alloc_count == 0);
+		prof_recent_alloc_head = node;
+		prof_recent_alloc_tail = node;
+	} else {
+		assert(prof_recent_alloc_tail != NULL);
+		assert(prof_recent_alloc_count > 0);
+		prof_recent_alloc_tail->next = node;
+		prof_recent_alloc_tail = node;
+	}
+	++prof_recent_alloc_count;
+	assert(prof_recent_alloc_max == -1 || (prof_recent_alloc_max > 0 &&
+	    prof_recent_alloc_count <= prof_recent_alloc_max));
+
+	malloc_mutex_unlock(tsd_tsdn(tsd), &prof_recent_alloc_mtx);
+
+	/*
+	 * Asynchronously handle the tctx of the head node, so that there's no
+	 * simultaneous holdings of prof_recent_alloc_mtx and tdata->lock.
+	 * In the worst case this may delay the release of the head_tctx but
+	 * it's better than holding prof_recent_alloc_mtx for longer.
+	 */
+	if (head_tctx != NULL) {
+		decrement_ref_count(tsd, head_tctx);
+	}
+	return false;
+}
+
+ssize_t
+prof_recent_alloc_max_get(tsdn_t *tsdn) {
+	ssize_t ret;
+	malloc_mutex_lock(tsdn, &prof_recent_alloc_mtx);
+	ret = prof_recent_alloc_max;
+	malloc_mutex_unlock(tsdn, &prof_recent_alloc_mtx);
+	return ret;
+}
+
+ssize_t
+prof_recent_alloc_max_set(tsd_t *tsd, ssize_t max) {
+	assert(max >= -1);
+
+	malloc_mutex_lock(tsd_tsdn(tsd), &prof_recent_alloc_mtx);
+	const ssize_t old_max = prof_recent_alloc_max;
+	prof_recent_alloc_max = max;
+
+	if (max == -1 || prof_recent_alloc_count <= max) {
+		/* Easy case - no need to alter the list. */
+		malloc_mutex_unlock(tsd_tsdn(tsd), &prof_recent_alloc_mtx);
+		return old_max;
+	}
+
+	prof_recent_alloc_t *begin = prof_recent_alloc_head;
+	do {
+		assert(prof_recent_alloc_head != NULL);
+		prof_recent_alloc_head = prof_recent_alloc_head->next;
```

Comment:
nit: we can avoid operating on the global data many times: have a local ptr to chase down the list N times, and then adjust prof_recent_alloc_head and prof_recent_alloc_count in one shot.